### PR TITLE
[Issue 527] fix send goroutine blocked

### PR DESCRIPTION
**Pull request by jiangbo9510**
Fix Issue: #527

### Motivation

I add a function to compare and poll block queue in atomic. To fix poll a wrong item in `failTimeoutMessages()`

### Modifications

This change is a trivial rework / code cleanup without any test coverage.

### Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)

### Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: pulsar/producer_partition.go [hunk 1]
```diff
@@ -490,21 +490,36 @@ func (p *partitionProducer) failTimeoutMessages() {
 
 		// iterate at most viewSize items
 		for i := 0; i < viewSize; i++ {
-			item := p.pendingQueue.Poll()
+			tickerNeedWaiting := time.Duration(0)
+			item := p.pendingQueue.CompareAndPoll(
```
Comment (merlimat) [inline pulsar/producer_partition.go:494]: This should also be used to refactor the `pendingQueue.Peek()` at line 464

Comment (ideadsnow) [reply pulsar/producer_partition.go:494]: the bug reason is peek head item and determine if it timeout, and then get a snapshot of pendingQueue, range snapshot and poll all the items in snapshot even if it is not all timeout.
the pendingQueue.Peek() at line 464 just peek(not poll) the head item in queue and check(logically correct), there are no actual operation to pendingQueue, so it looks like no need to refactor

**Comment by jiangbo9510** [comment]
@wolfstudy  When this pull request merge to the master branch?

**Comment by jiangbo9510** [comment]
@merlimat  please review my code and merge this pull request to master branch.

**Review by wolfstudy — APPROVED** [review]
LGTM +1

**Comment by wolfstudy** [comment]
cc / @cckellogg @merlimat  PTAL

**Comment by wolfstudy** [comment]
@jiangbo9510 Please merger master code for failed action CI

**Comment by jiangbo9510** [comment]
> @jiangbo9510 Please merger master code for failed action CI

What does this mean?
Should I merge master to my branch?

**Review by wolfstudy — CHANGES_REQUESTED** [review]
@jiangbo9510 This change may not really solve the problem

**Comment by wolfstudy** [comment]
@jiangbo9510 https://github.com/apache/pulsar-client-go/issues/565

**Comment by jiangbo9510** [comment]
> @jiangbo9510 #565

@wolfstudy  It's anothor problem. If it occured,All send goroutins will blocked by the dead lock.

In my case, Sometimes some send goroutins blocked. and others goroutins can send message success.

**Review by wolfstudy — APPROVED** [review]

**Pull request merged**